### ISSUE TITLE
Transpiled query complains about "unknown column" in not-equal predicate

### ISSUE THREAD

**Issue by wolffcm** [influxdata/influxdb#10741]
1.x InfluxQL will produce a result for something like this:
```
SELECT sum(f) 
FROM source 
WHERE non_existent_tag_key != 'value' 
    AND time >= now() - 1h 
GROUP BY tag1, tag2
```
Even though none of the series in the measurement have `non_existent_tag_key`, this produces a result.  I guess because the `non_existent_tag_key` predicate will evaluate to true for the `!=` operator.

However, when this query is transpiled and run through Flux, the following error occurs:
```
function references unknown column \"non_existent_tag\"
```

**Comment by wolffcm** [comment]
One way to handle this would be:
- Have a Flux function `valueIfExists(r: <record>, column: "somecol")` that returns a value if the field exists on a record, and `null` otherwise
- Have a Flux function called `isNull(v: value)` that returns `true` if a value is null and false otherwise
- Have some kind of conditional expression (I used C-style syntax in the example below, but I don't have a strong preference)
- Add functions to the Flux package `influxdata/influxdb/v1`:
```
fieldNotEqual = (r, column, rhs) => {
    v = valueIfExists(r: r, column: column)
    return  isNull(v: v) ? true : v != rhs
}
```
- Modify the transpiler to use `v1.fieldNotEqual` instead of vanilla `!=`.  Do something similar for the other kinds of operators as needed.

Some notes:
- `valueIfExists` could be partially evaluated if we know such a column exists via type inference, or it could be evaluated at runtime if we're not sure.
- We already have syntax that references columns of tables as strings (all the `columns` arguments of various functions).
- I have work on a branch from hackday that makes conditional expressions work in Flux.
- We would have to have `null` support for the scalar part of Flux working, but we would need to do that anyway.

**Comment by nathanielc** [comment]
What if the transpiler inspected the tags before generating the Flux code? This way the transpiler would see the tag is missing and know that part of the condition evals to `true`, in which case it can simplify the expression. This way Flux does not need to change. 

In a way you can think of the transpiler as  partial evaluation step where it can partially evaluate the where condition for these cases and generate safe Flux code.

**Comment by pauldix** [comment]
This issue might come up in other transpilers, like PromQL, so it would be nice to be able to represent this in a single Flux query.

**Comment by nathanielc** [comment]
@wolffcm that is a good point about it sometimes existing. I think this opens up to the higher level questions we had about how to balance the dynamic typing vs static typing that flexible schemas introduce. I like the idea of a function that partially evaluates away for most cases... 

Maybe we should leverage `null` here?  Meaning `null` represents a missing value, so reading a non existent column could be represented as null. 

I have had this vague idea brewing that we can leverage partial evaluation to bridge the gap between unknown types and known types. What follows is a slightly tangential braindump of some of my vague ideas on how to deal with messy data. The case of some series having tags and other not is a specific case of the larger issue of having messy data.

* Dynamic types allow for expressing flexible code that can handle different type of values. They also make it hard to write code because you can never be sure what the type really is. The code has to be written defensively around the type system.
* Static types provide information that can be leveraged to find errors before runtime and leverage for optimization. Code can eliminate type checks and make safe assumptions about the types of input data.
* Working with data is messy, having dynamic types allows for representing messy data. Traditionally the initial steps in working with data are to clean it up into a know format. This is where static types come in.

I wonder if there is a way to design Flux's type system to be partially dynamic and partially static. Meaning that at first we cannot assume any type information about the data but as we process it we can clean it up to be in a know state. This way the code that touches the raw data has to be written defensively while code afterwards can be written in a more straightforward manner. 

For example your idea of `valueIfExists` function is a function that takes input data that has dynamic type and outputs data with a known type. It serves as the bridge between the two worlds. 

We had a similar idea with type casts of columns. For example maybe a table can have a column of mixed type and the type cast function `float` takes any numeric type and converts it to a float type. It seems like the kind env in the type inference logic could represent this.

If we use a base type like Go's `interface{}` then everything is always typed (i.e. static typing) but we may not know anything about the type. Then using the kind env we can begin to put more constraints on what we know about the type. 

TL;DR  Flux becomes a language that can expression messy data and how to transform it to clean data and once its clean you can leverage a static type system to operate on it.

Thoughts?

**Comment by wolffcm** [comment]
@nathanielc This is all very interesting.

> Maybe we should leverage `null` here? Meaning `null` represents a missing value, so reading a non existent column could be represented as null.

I don't think that referencing a non-existent column should silently return `null`.  If a user misspells a column name, something should fail (preferably before execution), rather than assume they meant an ephemeral column that only contains `null` values.  

(I *might* be convinced that `r["bogus"]` should return null because the syntax is `map`-like.  But `r.bogus` looks like a C struct and should be an error for sure.  Currently both syntaxes are equivalent, I think?)

> TL;DR Flux becomes a language that can expression messy data and how to transform it to clean data and once its clean you can leverage a static type system to operate on it.

I like this approach.  Who is responsible for doing the clean-up of messy data, users, or Flux?  Maybe we provide helpful library functions so the user can avoid having to do the most common kinds of clean up.  Or maybe the compiler/planner instruments the query with sanitizing operations automatically.

From the perspective of the type system, what you are describing is pretty much how things work now, right?  The tables in a table stream have the type `{; any}`, with a lower bound that doesn't require any fields, and an upper bound that allows any field.  We don't refine table types at all yet, but I assume moving forward that we'll want to narrow the gap between the upper and lower bounds as much as we can.

At least in the transpiler, in order to be able to refine types we'd need to ask storage which tags exist.  This brings up some questions for me.  What if the transpiler determines that no series exist with the tag `bogus`, but someone adds points with this tag after planning but before execution.  The transpiled query would no longer be type-safe.  Do we care about this?  Do we just set `now` to be equal to the time that the transpiler issued the metaquery?

In a query engine for a language like SQL, columns can be added or dropped, but this should be infrequent.  If a query was planned against an old version of the schema, the plan can be discarded and the query re-planned.  But for InfluxDB, tag creation happens more frequently than schema changes do in SQL.

**Comment by wolffcm** [comment]
Been thinking more about this specific case.  We would only need two new Flux functions:
- `valueIfExists(r: <record>, column: "columnname")`
- `isNull(v: <value>)`

There is no need for a conditional expression or for null support other than these two functions.  To get relational operators that are equivalent to InfluxQL:
```
package v1

tagNotEquals = (r, column, rhs) => {
  v = valueIfExists(r: r, column: column)
  return isNull(c) or v != rhs
}

tagEquals = (r, column, rhs) => {
  v = valueIfExists(r: r, column: column)
  return not isNull(c) and v == rhs
}
```

**Comment by wolffcm** [comment]
Added related issues to implement `isNull` and `valueIfExists`:
https://github.com/influxdata/flux/issues/965
https://github.com/influxdata/flux/issues/966
